src/controllers/queries/index.ts

```diff
@@ -184,6 +203,114 @@ export class QueriesController extends HTMLController {
     )
   }
 
+  /**
+   * Retrieves the query response page
+   */
+  @SuccessResponse(200)
+  @Get('/scope-3-carbon-consumption/{queryId}/response')
+  public async scope3CarbonConsumptionResponse(@Path() queryId: UUID): Promise<HTML> {
+    this.logger.debug('query response page requested %j', { queryId })
+    const [query] = await this.db.get('query', { id: queryId })
+
+    if (!query) {
+      throw new NotFoundError(`There has been an issue retrieving the query.`)
+    }
+
+    const [connection] = await this.db.get('connection', { id: query.connection_id })
+    if (!connection) {
+      throw new InvalidInputError(`There has been an issue retrieving the connection.`)
+    }
+
+    return this.html(
+      this.scope3CarbonConsumptionResponseTemplates.newScope3CarbonConsumptionResponseFormPage(
+        'form',
+        connection,
+        queryId,
+        query.details['quantity'],
+        query.details['productId']
+      )
+    )
+  }
+
+  /**
+   * Submits the query response page
+   */
+  @SuccessResponse(200)
+  @Post('/scope-3-carbon-consumption/{queryId}/response')
+  public async scope3CarbonConsumptionResponseSubmit(
+    @Path() queryId: UUID,
+    @Body()
+    body: {
+      companyId: UUID
+      action: 'success'
+      totalScope3CarbonEmissions: string
+      partialResponse?: number
+    }
+  ): Promise<HTML> {
+    this.logger.debug('query page requested')
+
+    const [connection] = await this.db.get('connection', { id: body.companyId, status: 'verified_both' }, [
+      ['updated_at', 'desc'],
+    ])
+    if (!connection) {
+      throw new NotFoundError(`Invalid connection ${body.companyId}`)
+    }
+    if (!connection.agent_connection_id || connection.status !== 'verified_both') {
+      throw new NotFoundError(`Cannot query unverified connection`)
+    }
+    const [queryRow] = await this.db.get('query', { id: queryId })
+    if (!queryRow.response_id) {
+      throw new NotFoundError(`Missing queryId to respond to.`)
+    }
+    const query = {
+      emissions: body.totalScope3CarbonEmissions,
+      queryIdForResponse: queryRow.response_id,
+    }
+    //send a drpc message with response
+    let rpcResponse: DrpcResponse
+    try {
+      const maybeResponse = await this.cloudagent.submitDrpcRequest(
+        connection.agent_connection_id,
+        'submit_query_response',
+        {
+          query: 'Scope 3 Carbon Consumption',
+          ...query,
+        }
+      )
+      if (!maybeResponse) {
+        return await this.handleError(queryRow, connection)
+      }
+      rpcResponse = maybeResponse
+    } catch (err) {
+      return await this.handleError(queryRow, connection, undefined, err)
+    }
+    const { result, error, id: rpcId } = rpcResponse
+
+    await this.db.insert('query_rpc', {
```

Comment:
just wonder why it's not all under one `try/catch` block?

Reply:
I htink this is sensible. We expect the `submitDrpcRequest` call might fail and want to handle that specifically. Errors elsewhere would be unexpected and can be handled by the global handler

Reply:
yea, makes sense, otherwise would say we do have a middleware that should handle all exceptions and for unknown - return 5xx